In [6]:
import pandas as pd
import json
import ast
import numpy as np 

In [2]:
# Ajustar la opción para mostrar más caracteres en cada columna
pd.set_option('display.max_colwidth', None)

1 - Lectura de Archivos

In [3]:
# Abrir el archivo JSON y leer línea por línea
with open(r'C:\\Users\\Pablo\\OneDrive\\Henry\\ProyectoFinal\\Google Maps\\metadata-sitios\\1.json') as file:
    data = [json.loads(line) for line in file]

# Convertir la lista de diccionarios en un DataFrame
df1 = pd.DataFrame(data)

In [4]:
# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\2.json'
# Leer el JSON en un DataFrame
df2 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\3.json'
# Leer el JSON en un DataFrame
df3 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\4.json'
# Leer el JSON en un DataFrame
df4 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\5.json'
# Leer el JSON en un DataFrame
df5 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\6.json'
# Leer el JSON en un DataFrame
df6 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\7.json'
# Leer el JSON en un DataFrame
df7 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\8.json'
# Leer el JSON en un DataFrame
df8 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\9.json'
# Leer el JSON en un DataFrame
df9 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\10.json'
# Leer el JSON en un DataFrame
df10 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\11.json'
# Leer el JSON en un DataFrame
df11 = pd.read_json(ruta_json, lines=True)

2 - Funciones para el analisis

In [10]:
def extractCity(df):
    """
    Esta función toma un DataFrame que contiene una columna llamada 'address', 
    la cual tiene direcciones formateadas con comas. La función extrae el valor
    que está entre la segunda y tercera coma (que representa la ciudad) y lo 
    almacena en una nueva columna llamada 'city'.

    Parámetros:
    df (DataFrame): El DataFrame que contiene la columna 'address'.

    Retorna:
    DataFrame: El DataFrame original con una nueva columna 'city' que contiene 
               el valor entre la segunda y tercera coma de 'address'.
    """
    
    # Dividir la columna 'address' en partes usando la coma como delimitador
    # Luego, extraer el segundo elemento (índice 2), que está entre la segunda y tercera coma

    df['city'] = df['address'].str.split(',').str[2].str.strip()
    
    return df

In [25]:
def nulosYVacios(df):
    """
    Este procedimiento recibe un DataFrame y muestra la cantidad de valores nulos, vacíos y la cantidad de 
    valores duplicados en todas las columnas excepto en las 6, 10, 11 y 13 del DataFrame. También muestra 
    el mínimo, máximo y media para columnas de tipo float e integer.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    print("Informe de Nulos, Vacíos y Duplicados:")
    print("-" * 30)

    # Total de registros en el DataFrame
    total_registros = len(df)

    # Imprimir total de registros
    print(f"Total de registros: {total_registros}")
    print("-" * 30)

    # Iterar sobre todas las columnas para mostrar nulos y vacíos
    for columna in df.columns:
        # Contar valores nulos
        nulos = df[columna].isnull().sum()

        # Contar valores vacíos (considerando que solo se cuentan si son cadenas)
        vacios = df[columna].apply(lambda x: isinstance(x, str) and x == '').sum()

        # Calcular porcentajes
        porcentaje_nulos = (nulos / total_registros) * 100 if total_registros > 0 else 0
        porcentaje_vacios = (vacios / total_registros) * 100 if total_registros > 0 else 0

        # Imprimir resultados
        print(f"Columna: {columna}")
        print(f"  Valores Nulos: {nulos}")
        print(f"  Valores Vacíos: {vacios}")
        print(f"  Porcentaje Nulos: {porcentaje_nulos:.2f}%")
        print(f"  Porcentaje Vacíos: {porcentaje_vacios:.2f}%")

        # Contar y mostrar la cantidad de valores duplicados solo si no es una de las columnas con datos anidados
        if columna not in [df.columns[6], df.columns[10], df.columns[11], df.columns[13]]:
            # Filtrar los valores no nulos antes de contar duplicados
            cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
            print(f"  Cantidad de valores duplicados (solo no nulos): {cantidad_duplicados}")

        # Mostrar mínimo, máximo y media si la columna es de tipo float o int
        if df[columna].dtype in ['float64', 'int64']:
            minimo = df[columna].min()
            maximo = df[columna].max()
            media = df[columna].mean()
            print(f"  Mínimo: {minimo}")
            print(f"  Máximo: {maximo}")
            print(f"  Media: {media:.2f}")

        print("-" * 30)


In [18]:
def rankingFrecuenciaCiudad(df, x=10):
    """
    Este procedimiento recibe un DataFrame y un número entero x.
    Muestra en pantalla el ranking de los primeros x elementos con mayor frecuencia en la columna 'city'.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    x (int): El número de elementos a mostrar en el ranking.
    """
    print(f"Ranking de los primeros {x} elementos más frecuentes en la columna 'city':")
    print("-" * 60)

    # Verifica si la columna 'city' existe en el DataFrame
    if 'city' in df.columns:
        # Obtener la frecuencia de cada valor en la columna 'city'
        frecuencias = df['city'].value_counts()

        # Obtener los primeros x elementos con mayor frecuencia
        ranking = frecuencias.head(x)

        # Mostrar el ranking
        for index, value in ranking.items():
            print(f"{index}: {value}")
    else:
        print("La columna 'city' no existe en el DataFrame.")

    print("-" * 60)

In [17]:
def mostrarCantidadFilasDuplicadas(df):
    """
    Filtra las filas duplicadas en un DataFrame según columnas específicas,
    y muestra la cantidad de filas duplicadas en la salida de pantalla.
  
    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    # Filtrar duplicados considerando las columnas específicas
    filas_duplicadas = df[df.duplicated(subset=[df.columns[i] for i in [0, 1, 2, 4, 5]], keep=False)]
    
    # Contar la cantidad de filas duplicadas
    cantidad_duplicados = len(filas_duplicadas)
    
    # Mostrar la cantidad de filas duplicadas
    print(f"Cantidad de filas duplicadas: {cantidad_duplicados}")

In [16]:
def encontrarDuplicados(df, columnas_excluir=None):
    """
    Encuentra filas duplicadas en un DataFrame basándose en todas las columnas, 
    excepto algunas (las que tienen datos anidados) que pueden ser excluidas.
    
    Args:
        df (pd.DataFrame): El DataFrame donde se buscarán duplicados.
        columnas_excluir (list, optional): Lista de nombres o índices de columnas a excluir del chequeo 
                                           de duplicados. Si es None, se excluyen las columnas 6, 10, 11 y 13.
        
    Returns:
        pd.DataFrame: Un DataFrame que contiene solo las filas duplicadas. 
    """
    # Definir columnas a excluir por defecto si no se proporcionan
    if columnas_excluir is None:
        columnas_excluir = [6, 10, 11, 13]
    
    # Crear una lista de todas las columnas menos las excluidas
    columnas_incluir = df.columns.difference(df.columns[columnas_excluir])
    
    # Verificar duplicados solo en las columnas seleccionadas
    duplicados = df.duplicated(subset=columnas_incluir, keep=False)
    
    # Retornar filas duplicadas
    return df[duplicados]


In [12]:
def eliminarDuplicados(df, columnas_excluir=None):
    """
    Crea una copia del DataFrame excluyendo solo las filas duplicadas, conservando la primera aparición de cada fila duplicada.
    
    Args:
        df (pd.DataFrame): El DataFrame original.
        columnas_excluir (list, optional): Lista de nombres o índices de columnas a excluir del chequeo 
                                           de duplicados. Si es None, se excluyen las columnas 6, 10, 11 y 13.
        
    Returns:
        pd.DataFrame: Una copia del DataFrame sin las filas duplicadas (se conserva la primera aparición).
    """
    # Definir las columnas a excluir por defecto si no se proporcionan
    if columnas_excluir is None:
        columnas_excluir = [6, 10, 11, 13]
    
    # Crear una lista de todas las columnas menos las excluidas
    columnas_incluir = df.columns.difference(df.columns[columnas_excluir])
    
    # Identificar las filas duplicadas, pero conservando la primera aparición de cada duplicado
    duplicados = df.duplicated(subset=columnas_incluir, keep='first')
    
    # Crear una copia del DataFrame excluyendo las filas duplicadas (segundas apariciones en adelante)
    df_sin_duplicados = df[~duplicados]
    
    return df_sin_duplicados


In [16]:
def desanidarColumna(df, col_index):
    """
    Explota la columna especificada por su índice en el DataFrame 
    y mantiene la columna 'gmap_id' en el nuevo DataFrame. 
    También elimina la columna explotada del DataFrame original.
    
    Args:
    df (pd.DataFrame): DataFrame que contiene la columna a desanidar.
    col_index (int): Índice de la columna a desanidar.
    
    Returns:
    tuple: (pd.DataFrame: DataFrame original actualizado, pd.DataFrame: Nuevo DataFrame con 'gmap_id' y la columna explotada)
    """
    # Obtener el nombre de la columna a partir de su índice
    col_name = df.columns[col_index]
    
    # Crear un nuevo DataFrame desanidado con explode
    df_exploded = df.explode(col_name).reset_index(drop=True)
    
    # Seleccionar solo las columnas 'gmap_id' y la columna explotada
    df_result = df_exploded[['gmap_id', col_name]]
    
    # Eliminar la columna especificada del DataFrame original
    df.drop(columns=[col_name], inplace=True)
    
    return df, df_result


Frecuencia de recomendaciones por ciudad segun google Maps

Unir todos los Data Frames:

In [12]:
# Lista de DataFrames a unir:
dataframes = [df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11]

# Unir todos los DataFrames
df_unido = pd.concat(dataframes, axis=0)

# Guardar el DataFrame unido como archivo Parquet
df_unido.to_parquet('df_unido.parquet')


In [7]:
# Leer archivo parquet
df_unido = pd.read_parquet(r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\df_unido.parquet')

In [11]:
# Añadir la ciudad a la que pertenecen los comercios
extractCity(df_unido)

name  \
0                      Porter Pharmacy   
1                         City Textile   
2                         San Soo Dang   
3                         Nova Fabrics   
4                     Nobel Textile Co   
...                                ...   
274996                        Steak 48   
274997   Jack Mcnerney Chevrolet, Inc.   
274998  Central Ny Spay Neuter Assista   
274999                   Ok Feed Store   
275000              Crestview Crossing   

                                                  address  \
0       Porter Pharmacy, 129 N Second St, Cochran, GA ...   
1       City Textile, 3001 E Pico Blvd, Los Angeles, C...   
2       San Soo Dang, 761 S Vermont Ave, Los Angeles, ...   
3       Nova Fabrics, 2200 E 11th St, Los Angeles, CA ...   
4       Nobel Textile Co, 719 E 9th St, Los Angeles, C...   
...                                                   ...   
274996   Steak 48, 260 S Broad St, Philadelphia, PA 19102   
274997  Jack Mcnerney Chevrolet, Inc., 363 NY-281, Tul...   
274998  Central Ny Spay Neuter Assista, 17 Salisbury S...   
274999  Ok Feed Store, 16300 SW 296th St, Homestead, F...   
275000              Crestview Crossing, Newberg, OR 97132   

                                      gmap_id description   latitude  \
0       0x88f16e41928ff687:0x883dad4fd048e8f8        None  32.388300   
1       0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None  34.018891   
2       0x80c2c778e3b73d33:0xbdc58662a4a97d49        None  34.058092   
3        0x80c2c89923b27a41:0x32041559418d447        None  34.023669   
4       0x80c2c632f933b073:0xc31785961fe826a6        None  34.036694   
...                                       ...         ...        ...   
274996  0x89c6c74f43a49b55:0x6be6995921c58b12        None  39.947254   
274997  0x89da1787d3fdc4a5:0xce92ea6e3cd4d1cc        None  42.788636   
274998  0x89da6b9ce832726f:0xbd0d3630f1e3b42e        None  42.603831   
274999  0x88d9e872a9e39f3d:0x5de29f50507e5d73        None  25.491641   
275000  0x54956b3944b1654d:0x7797cead9a485c22        None  45.310254   

         longitude                                           category  \
0       -83.357100                                         [Pharmacy]   
1      -118.215290                                 [Textile exporter]   
2      -118.292130                                [Korean restaurant]   
3      -118.232930                                     [Fabric store]   
4      -118.249421                                     [Fabric store]   
...            ...                                                ...   
274996  -75.164953  [Steak house, Fine dining restaurant, Seafood ...   
274997  -76.122120                                 [Chevrolet dealer]   
274998  -76.165286                                     [Veterinarian]   
274999  -80.456076                                [Animal feed store]   
275000 -122.935500                                          [Lodging]   

        avg_rating  num_of_reviews price  \
0              4.9              16  None   
1              4.5               6  None   
2              4.4              18  None   
3              3.3               6  None   
4              4.3               7  None   
...            ...             ...   ...   
274996         4.6             308  None   
274997         4.4             143  None   
274998         4.3              88  None   
274999         4.9              13  None   
275000         5.0               1  None   

                                                    hours  \
0       [[Friday, 8AM–6PM], [Saturday, 8AM–12PM], [Sun...   
1                                                    None   
2       [[Thursday, 6:30AM–6PM], [Friday, 6:30AM–6PM],...   
3       [[Thursday, 9AM–5PM], [Friday, 9AM–5PM], [Satu...   
4       [[Thursday, 9AM–5PM], [Friday, 9AM–5PM], [Satu...   
...                                                   ...   
274996  [[Monday, 4–10PM], [Tuesday, 4–10PM], [Wednesd...   
274997  [[Monday, 8:30AM–7PM], [Tuesd

In [27]:
# Ranking de Ciudades con mas reseñas
rankingFrecuenciaCiudad(df_unido)

Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 29669
New York: 20879
Chicago: 19850
Los Angeles: 17296
Brooklyn: 17186
San Antonio: 14742
Dallas: 13518
Las Vegas: 13329
Miami: 13130
Philadelphia: 12417
------------------------------------------------------------


In [26]:
# Obtencion de Datos para el analisis preliminar de calidad por columna
nulosYVacios(df_unido)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 3025011
------------------------------
Columna: name
  Valores Nulos: 37
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 870876
------------------------------
Columna: address
  Valores Nulos: 80511
  Valores Vacíos: 0
  Porcentaje Nulos: 2.66%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 57210
------------------------------
Columna: gmap_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 53156
------------------------------
Columna: description
  Valores Nulos: 2770722
  Valores Vacíos: 0
  Porcentaje Nulos: 91.59%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 187002
------------------------------
Columna: latitude
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nu

3 - Analisis Preeliminar de datos, primeros hallazgos:
Este informe proporciona un análisis exhaustivo de la calidad de los datos en un DataFrame de 3,025,011 registros. Aquí hay un resumen de los hallazgos más relevantes:

1. **Valores Nulos y Vacíos**:
   - Varias columnas, como `description`, `price`, y `hours`, presentan un alto número de valores nulos. En particular, la columna `description` tiene un 91.59% de valores nulos, lo que indica que la mayoría de las entradas no tienen descripción. En concecuencia estos datos habra que analizar su utilidad en el analisis.
   - La columna `price` también tiene un alto porcentaje de valores nulos (90.90%).

2. **Duplicados**:
   - La columna `gmap_id` tiene 53,156 valores duplicados, lo que sugiere que hay múltiples registros asociados con la misma identificación de Google Maps. Este tipo de duplicados podría ser un área de preocupación si se necesita unicidad en la identificación de lugares.

3. **Columnas Numéricas**:
   - Las columnas `latitude` y `longitude` tienen un rango amplio de valores, con latitudes que van desde -40.93 hasta 87.86 y longitudes desde -178.81 hasta 180.0. Esto puede ser esperado, pero es importante validar que todos los registros caen dentro de las coordenadas geográficas válidas.
   - La columna `avg_rating` tiene una media de 4.30, lo que sugiere una percepción mayormente positiva en la evaluación de los lugares.

4. **Valores Vacíos**:
   - Hay una cantidad mínima de valores vacíos en las columnas, lo que indica que la mayoría de las entradas están completas en este aspecto. Sin embargo, la columna `city` tiene 687 valores vacíos, lo que podría ser relevante si se realizan análisis geográficos.

5. **Observaciones Generales**:
   - Alta proporción de valores nulos en columnas clave como `description`, `price`, y `MISC`.



In [13]:
# Se procede a eliminar las filas que se encuentran repetidas, considerando el mismo creterio mencionado previamente.
dfLimpio = eliminarDuplicados(df_unido)


In [33]:
nulosYVacios(dfLimpio)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 2998428
------------------------------
Columna: name
  Valores Nulos: 37
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 823945
------------------------------
Columna: address
  Valores Nulos: 79520
  Valores Vacíos: 0
  Porcentaje Nulos: 2.65%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 6106
------------------------------
Columna: gmap_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 0
------------------------------
Columna: description
  Valores Nulos: 2745412
  Valores Vacíos: 0
  Porcentaje Nulos: 91.56%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 185439
------------------------------
Columna: latitude
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 

Comparando con el informe anterior, las principales diferencias son:

1. **Total de Registros**: El número total de registros ha disminuido de **3,025,011** a **2,998,428**.

2. **Valores Nulos en `name`**: Los valores nulos en la columna `name` han disminuido de **37** a **0**, lo que indica que todos los registros ahora tienen un nombre.

3. **Valores Nulos en `address`**: La cantidad de valores nulos en `address` ha aumentado ligeramente, de **80,511** a **79,520**, lo que representa un cambio mínimo en la calidad de esta columna.

4. **Valores Nulos en `description`**: El porcentaje de valores nulos en `description` se mantiene alto, con **2,745,412** valores nulos, lo que equivale al **91.56%** de los registros, sin cambios significativos.

5. **Valores Nulos en `price`**: El número de valores nulos en `price` ha aumentado, de **2,749,808** a **2,724,510**, lo que indica que todavía hay un alto porcentaje de datos faltantes.

6. **Duplicados en `address`**: La cantidad de duplicados en `address` ha aumentado, de **5,106** a **6,106**, lo que sugiere que a pesar de la eliminación de registros, todavía hay duplicados en esta columna.

7. **Duplicados en `gmap_id`**: La columna `gmap_id` ahora muestra **0** duplicados, indicando que se ha mantenido la unicidad en esta columna, lo cual es un punto positivo en comparación con la versión anterior.

8. **Valores Nulos en `state`**: El número de valores nulos en `state` ha disminuido, de **746,455** a **738,858**, lo que indica una ligera mejora en la completitud de esta columna.

En resumen, el nuevo informe muestra una ligera mejora en algunas columnas (como `name` y `gmap_id`), pero persisten problemas significativos de nulos y duplicados en columnas clave, lo que requiere atención adicional.

Desanidar Columnas que contienen lista:

In [14]:
dfPrueba = dfLimpio.copy()

In [17]:
dfPrueba, dfCategory = desanidarColumna(dfPrueba, 6)

In [64]:
# Guardar en un archivo .parquet
dfCategory.to_parquet('dfCategory.parquet', index=False)

In [62]:
dfPrueba.info()


<class 'pandas.core.frame.DataFrame'>
Index: 2998428 entries, 0 to 275000
Data columns (total 15 columns):
 #   Column            Dtype  
---  ------            -----  
 0   name              object 
 1   address           object 
 2   gmap_id           object 
 3   description       object 
 4   latitude          float64
 5   longitude         float64
 6   avg_rating        float64
 7   num_of_reviews    int64  
 8   price             object 
 9   hours             object 
 10  MISC              object 
 11  state             object 
 12  relative_results  object 
 13  url               object 
 14  city              object 
dtypes: float64(3), int64(1), object(11)
memory usage: 398.3+ MB


In [18]:
dfPrueba, dfRelativeResults = desanidarColumna(dfPrueba, 12)


In [65]:
# Guardar en un archivo .parquet
dfRelativeResults.to_parquet('dfRelativeResults.parquet', index=False)

In [21]:
dfPrueba.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2998428 entries, 0 to 275000
Data columns (total 14 columns):
 #   Column          Dtype  
---  ------          -----  
 0   name            object 
 1   address         object 
 2   gmap_id         object 
 3   description     object 
 4   latitude        float64
 5   longitude       float64
 6   avg_rating      float64
 7   num_of_reviews  int64  
 8   price           object 
 9   hours           object 
 10  MISC            object 
 11  state           object 
 12  url             object 
 13  city            object 
dtypes: float64(3), int64(1), object(10)
memory usage: 343.1+ MB


In [20]:
# Guardar como parquet archivo limpio sin Normalizar
dfPrueba.to_parquet('dfPrueba.parquet', index=False)

In [ ]:
# Copiar el DataFrame con las columnas 'gmap_id' y 'hours'
df_copy = dfPrueba[['gmap_id', 'hours']].copy()

# Reemplazar None con una lista vacía
df_copy['hours'] = df_copy['hours'].apply(lambda x: [] if x is None else x)

# Crear una lista de días de la semana
days_of_week = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

# Crear un DataFrame vacío para los horarios normalizados
hours_normalized = pd.DataFrame(index=df_copy.index)

# Iterar sobre los días de la semana y llenar el DataFrame normalizado
for day in days_of_week:
    hours_normalized[day] = df_copy['hours'].apply(
        lambda x: next((h[1] for h in x if h[0] == day), None)
    )

# Concatenar el DataFrame original con las columnas normalizadas
df_normalized = pd.concat([df_copy[['gmap_id']], hours_normalized], axis=1)

# Mostrar el nuevo DataFrame normalizado
print(df_normalized.head())

In [ ]:
# Concatenar dfPrueba y df_normalized por la columna gmap_id
df_final = pd.merge(dfPrueba, df_normalized, on='gmap_id', how='left')

# Eliminar la columna 'hours'
df_final = df_final.drop('hours', axis=1)

In [23]:
# Normalizar la columna 'MISC'
misc_normalized = pd.json_normalize(df_final['MISC'])

# Concatenar el DataFrame original (sin la columna MISC) con el nuevo DataFrame normalizado
df_normalizedMISC = pd.concat([df_final.drop(columns=['MISC']), misc_normalized], axis=1)



0         {'Accessibility': ['Wheelchair accessible entr...
1                                                      None
2         {'Accessibility': ['Wheelchair accessible entr...
3         {'Accessibility': None, 'Activities': None, 'A...
4         {'Accessibility': None, 'Activities': None, 'A...
                                ...                        
274996    {'Accessibility': ['Wheelchair accessible entr...
274997    {'Accessibility': None, 'Activities': None, 'A...
274998    {'Accessibility': None, 'Activities': None, 'A...
274999    {'Accessibility': None, 'Activities': None, 'A...
275000    {'Accessibility': ['Wheelchair accessible entr...
Name: MISC, Length: 2998428, dtype: object

Lectura de Archivos Normalizados con Colab

In [28]:
df_hours = pd.read_parquet(r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\DFnormalizados\df_normalized.parquet')

In [29]:
df_hours.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998428 entries, 0 to 2998427
Data columns (total 8 columns):
 #   Column     Dtype 
---  ------     ----- 
 0   gmap_id    object
 1   Monday     object
 2   Tuesday    object
 3   Wednesday  object
 4   Thursday   object
 5   Friday     object
 6   Saturday   object
 7   Sunday     object
dtypes: object(8)
memory usage: 183.0+ MB


In [30]:
df_normalized = pd.read_parquet(r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\DFnormalizados\dfNormalizadoMISC.parquet')

In [31]:
df_normalized.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998428 entries, 0 to 2998427
Data columns (total 31 columns):
 #   Column             Dtype  
---  ------             -----  
 0   name               object 
 1   address            object 
 2   gmap_id            object 
 3   description        object 
 4   latitude           float64
 5   longitude          float64
 6   avg_rating         float64
 7   num_of_reviews     int64  
 8   price              object 
 9   hours              object 
 10  state              object 
 11  url                object 
 12  city               object 
 13  Accessibility      object 
 14  Activities         object 
 15  Amenities          object 
 16  Atmosphere         object 
 17  Crowd              object 
 18  Dining options     object 
 19  From the business  object 
 20  Getting here       object 
 21  Health & safety    object 
 22  Health and safety  object 
 23  Highlights         object 
 24  Lodging options    object 
 25  Offerings         

In [32]:
# Eliminar la columna 'hours' del DataFrame df_final_normalized
df_normalized.drop(columns=['hours'], inplace=True)

In [33]:
# Unir df_final_normalized y df_hours por la columna 'gmap_id'
df_normalized = pd.merge(df_normalized, df_hours, on='gmap_id', how='left')

In [34]:
df_normalized.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998428 entries, 0 to 2998427
Data columns (total 37 columns):
 #   Column             Dtype  
---  ------             -----  
 0   name               object 
 1   address            object 
 2   gmap_id            object 
 3   description        object 
 4   latitude           float64
 5   longitude          float64
 6   avg_rating         float64
 7   num_of_reviews     int64  
 8   price              object 
 9   state              object 
 10  url                object 
 11  city               object 
 12  Accessibility      object 
 13  Activities         object 
 14  Amenities          object 
 15  Atmosphere         object 
 16  Crowd              object 
 17  Dining options     object 
 18  From the business  object 
 19  Getting here       object 
 20  Health & safety    object 
 21  Health and safety  object 
 22  Highlights         object 
 23  Lodging options    object 
 24  Offerings          object 
 25  Payments          

Resultado Final:
dfNormalizado.parquet       DF Union de los 11 json, sin filas duplicadas.
dfCategory.parquet          DF expandido de category, valor llave  gmap_id (columna eliminada)
dfRelativeResults.parquet   DF expandido de relative_results, valor llave  gmap_id (columna eliminada)

Data frames intermedios: 
df_unido.parquet
dfPrueba.parquet

In [36]:
df_normalized.to_parquet('dfNormalizado.parquet')

In [135]:
# Cargar el archivo de business
dfYeplp = pd.read_pickle(r'C:\Users\Pablo\OneDrive\Henry\Proyecto Final\Yelp\business.pkl')

In [152]:
dfYeplp.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 28 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
 14  business_id   5 non-null       object
 15  name          5 non-null       object
 16  address       5 non-null       object
 17  city          5 non-null       object
 18  state         5 non-null     

In [ ]:
# Filtrar las filas donde gmap_id está duplicado
filas_duplicadas = df1[df1['gmap_id'].duplicated(keep=False)]


In [ ]:
# Ordenar el DataFrame por la columna gmap_id
df_ordenado = filas_duplicadas.sort_values(by='gmap_id')


df_ordenado


In [36]:
df1.iloc[:,6]

0                                                                                                                                                       [Pharmacy]
1                                                                                                                                               [Textile exporter]
2                                                                                                                                              [Korean restaurant]
3                                                                                                                                                   [Fabric store]
4                                                                                                                                                   [Fabric store]
                                                                                    ...                                                                           
274996                

In [153]:
# Crear un nuevo DataFrame usando loc
dfYeplp_copia = dfYeplp.copy()

Eliminar columnas duplicadas, desde la 14 a 27

In [154]:
# Crear copia donde solo se excluyen las duplicadas.
dfYeplp_copia = dfYeplp_copia.iloc[:,0:14]

In [155]:
dfYeplp_copia.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
dtypes: object(14)
memory usage: 17.2+ MB


Ranking de ciudades con mas comentarios segun Yelp

In [1]:
rankingFrecuenciaCiudad(dfYeplp_copia, 5)

NameError: name 'rankingFrecuenciaCiudad' is not defined

In [8]:
# Eliminar las columnas 'url', 'relative_results', 'state' y 'price'
df_cleaned = df1.drop(columns=['url', 'relative_results', 'state', 'price'])

# Ver las primeras filas del DataFrame limpio
print(df_cleaned.head())

               name                                            address  \
0   Porter Pharmacy  Porter Pharmacy, 129 N Second St, Cochran, GA ...   
1      City Textile  City Textile, 3001 E Pico Blvd, Los Angeles, C...   
2      San Soo Dang  San Soo Dang, 761 S Vermont Ave, Los Angeles, ...   
3      Nova Fabrics  Nova Fabrics, 2200 E 11th St, Los Angeles, CA ...   
4  Nobel Textile Co  Nobel Textile Co, 719 E 9th St, Los Angeles, C...   

                                 gmap_id description   latitude   longitude  \
0  0x88f16e41928ff687:0x883dad4fd048e8f8        None  32.388300  -83.357100   
1  0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None  34.018891 -118.215290   
2  0x80c2c778e3b73d33:0xbdc58662a4a97d49        None  34.058092 -118.292130   
3   0x80c2c89923b27a41:0x32041559418d447        None  34.023669 -118.232930   
4  0x80c2c632f933b073:0xc31785961fe826a6        None  34.036694 -118.249421   

              category  avg_rating  num_of_reviews  \
0           [Pharmacy]    

In [13]:
print(df_cleaned['MISC'][0])


{'Service options': ['In-store shopping', 'Same-day delivery'], 'Health & safety': ['Mask required', 'Staff required to disinfect surfaces between visits'], 'Accessibility': ['Wheelchair accessible entrance'], 'Planning': ['Quick visit']}


In [19]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275001 entries, 0 to 275000
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   name            274994 non-null  object 
 1   address         264939 non-null  object 
 2   gmap_id         275001 non-null  object 
 3   description     13155 non-null   object 
 4   latitude        275001 non-null  float64
 5   longitude       275001 non-null  float64
 6   category        272740 non-null  object 
 7   avg_rating      275001 non-null  float64
 8   num_of_reviews  275001 non-null  int64  
 9   hours           192448 non-null  object 
 10  MISC            194972 non-null  object 
dtypes: float64(3), int64(1), object(7)
memory usage: 23.1+ MB
